how-to-guides/time-profiling.ipynb

```diff
@@ -0,0 +1,939 @@
+      },
+      "source": [
+        "trainer.run(train_loader, max_epochs=2)\n",
+        "\n",
+        "print(\n",
+        "    f\"Total time: {timer.total:.2f}, Average time taken by a single epoch: {timer.value():.2f}\"\n",
+        ")"
+      ],
+      "execution_count": 51,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "text": [
+            "Training Results - Epoch[1] Avg accuracy: 0.97 Avg loss: 0.11\n",
+            "Validation Results - Epoch[1] Avg accuracy: 0.97 Avg loss: 0.10\n",
+            "Training Results - Epoch[2] Avg accuracy: 0.99 Avg loss: 0.05\n",
+            "Validation Results - Epoch[2] Avg accuracy: 0.99 Avg loss: 0.04\n",
+            "Total time: 93.11, Average time taken by a single epoch: 46.55\n"
+          ],
+          "name": "stdout"
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "kTD4UFuZWUSe"
+      },
+      "source": [
+        "## Using `State` of Events\n",
+        "\n",
+        "If we just want to print the time taken after every epoch and the total time for training we can simply use the `trainer`'s [`State`](https://pytorch.org/ignite/generated/ignite.engine.events.State.html#ignite.engine.events.State). We attach two separate handlers fired when an epoch is completed and when the training is completed to log the time returned by `trainer.state.times`."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "2m1pOALZ3vCO"
+      },
+      "source": [
+        "@trainer.on(Events.EPOCH_COMPLETED)\n",
+        "def log_epoch_time():\n",
+        "    print(\n",
+        "        f\"Epoch {trainer.state.epoch}, Time Taken : {trainer.state.times['EPOCH_COMPLETED']}\"\n",
+        "    )\n",
+        "\n",
+        "\n",
+        "@trainer.on(Events.COMPLETED)\n",
+        "def log_total_time():\n",
+        "    print(f\"Total Time: {trainer.state.times['COMPLETED']}\")"
+      ],
+      "execution_count": 52,
+      "outputs": []
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "hmdHZIDIABc9",
+        "outputId": "eaa53283-3e7f-4cdf-a013-f7fde3baed83"
+      },
+      "source": [
+        "trainer.run(train_loader, max_epochs=2)"
+      ],
+      "execution_count": 53,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "text": [
+            "Training Results - Epoch[1] Avg accuracy: 0.99 Avg loss: 0.03\n",
+            "Validation Results - Epoch[1] Avg accuracy: 0.99 Avg loss: 0.03\n",
+            "Epoch 1, Time Taken : 31.631677865982056\n",
+            "Training Results - Epoch[2] Avg accuracy: 0.99 Avg loss: 0.03\n",
+            "Validation Results - Epoch[2] Avg accuracy: 0.99 Avg loss: 0.04\n",
+            "Epoch 2, Time Taken : 32.099266052246094\n",
+            "Total Time: 93.50223469734192\n"
```

Comment:
Total time is almost 93s and epoch 30s. It’s quite different to the previous example using `Timer`, right ?

Reply:
Thank you for pointing that out! I just noticed that and realized that the results from `Timer()` also included the time taken by `log_training_results()` and `log_validation_results()`. I presume it's because those two handlers were attached before `Timer()`. I will add an explanation of why this happens.